### PR TITLE
`LocaleFallbackerBorrowed` and `DataKey`<->`LocaleFallbackConfig` conversions

### PR DESCRIPTION
**Pull request by robertbastian**
Currently the borrowed form of `LocaleFallbacker` is `LocaleFallbackerWithConfig`, which borrows `LocaleFallbacker`'s payloads and adds an owned `LocaleFallbackConfig`, which is then borrowed by `LocaleFallbackIterator`. This is overly verbose, because using this always requires defining a variable for the intermediate `LocaleFallbackerWithConfig`, and that cannot be reused between keys.

### CODE REVIEW

**Review by Manishearth — DISMISSED** [review]
seems fine, probably should also get review from shane

**Comment by sffc** [comment]
One set of moves that would work:

- Keep the enums where they are
- Re-export the enums from icu_locid_transform
- Move everything except the Adapter into icu_locid_transform, including public APIs involving DataKey

**Comment by robertbastian** [comment]
Force pushed because I reverted a lot

**Review by Manishearth — APPROVED** [review]

File: provider/adapters/src/fallback/mod.rs [hunk 1]
```diff
@@ -309,72 +336,64 @@ impl LocaleFallbacker {
...
+    /// When first initialized, the locale is normalized according to the fallback algorithm.
+    ///
+    /// For performance reasons, internally an [`icu_provider::DataLocale`] is used.
+    ///
+    /// [`Locale`]: icu_locid::Locale
+    #[doc(hidden)] // for now
+    pub fn fallback_for(
+        &self,
+        config: LocaleFallbackConfig,
+        locale: impl Into<DataLocale>,
```
Comment (sffc) [inline provider/adapters/src/fallback/mod.rs:350]: Nit (optional): Still not the hugest fan of `impl Into` arguments; just let people write `.into()` at the call site for the same effect and no need to import extra types

**Review by sffc — APPROVED** [review]

**Pull request merged**